laplace/curvature/asdl.py

```diff
@@ -107,33 +128,81 @@ def _rescale_kron_factors(kron, N):
                 F[1] *= 1/N
         return kron
 
-    def diag(self, X, y, **kwargs):
-        with torch.no_grad():
-            if self.last_layer:
-                f, X = self.model.forward_with_features(X)
-            else:
-                f = self.model(X)
-            loss = self.lossfunc(f, y)
-        curv = fisher_for_cross_entropy(self._model, self._ggn_type, SHAPE_DIAG,
-                                        inputs=X, targets=y)
-        diag_ggn = curv.matrices_to_vector(None)
+    def diag(self, X, y, N=None, **kwargs):
+        del N
+        if self.last_layer:
+            _, X = self.model.forward_with_features(X)
+        cfg = FisherConfig(fisher_type=self._ggn_type, loss_type=self.loss_type,
+                           fisher_shapes=[SHAPE_DIAG], data_size=1, **kwargs)
+        fisher_maker = get_fisher_maker(self.model, cfg)
+        y = y if self.loss_type == LOSS_MSE else y.view(-1)
+        if 'emp' in self._ggn_type:
+            dummy = fisher_maker.setup_model_call(self._model, X)
+            dummy = dummy if self.loss_type == LOSS_MSE else dummy.view(-1, dummy.size(-1))
+            fisher_maker.setup_loss_call(self.lossfunc, dummy, y)
+        else:
+            fisher_maker.setup_model_call(self._model, X)
+        f, _ = fisher_maker.forward_and_backward()
+        # Assumes that the last dimension of f is of size outputs.
+        f = f if self.loss_type == LOSS_MSE else f.view(-1, f.size(-1))
```

Comment:
This is against the convention used for `CrossEntropyLoss` in PyTorch, since it expects input shape `(N, C, d_1​, d_2 ​,..., d_K​)` (see the [docs](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)). Could lead to unexpected behaviour for users. If we go with this convention, we should comment on it in the docs and the README, as a requirement for the model.

Reply:
It's from ASDL: https://github.com/kazukiosawa/asdl/blob/9ba6a54bfa4cf3fb21dbc2922f6ef128a3b2e0f9/asdl/fisher.py#L456-L457

Does you suggest we also modify ASDL?

Reply:
After some readings, I think the current implementation is fine. The input of the model is `N, d_1, ..., d_K, D` where `d_1, ..., d_K` are additional "batch" dimensions and `D` is the input-space dim. Then the model will always output `N, d_1, ..., d_K, C` where `C` is the number of classes.

In Huggingface, they always assume that the logits and labels are flattended, e.g. here for [GPT2](https://github.com/huggingface/transformers/blob/b109257f4fb8b1166e7c53cc5418632014ed53a5/src/transformers/models/gpt2/modeling_gpt2.py#L1336-L1338) and [LLAMA-2](https://github.com/huggingface/transformers/blob/b109257f4fb8b1166e7c53cc5418632014ed53a5/src/transformers/models/llama/modeling_llama.py#L1218-L1224). The loss is then oblivious and agnostic to the fact that we have additional batch dimensions. This is also the current assumption of Laplace and ASDL.

So, I believe as long as we consistently follow the flatten-and-compute-loss convention, we're clear. The user won't be affected at all since it's an internal implementation.

Reply:
Looking forward to hear your thoughts, though, since I might be wrong!